nsqd/nsqd.go

```diff
@@ -551,6 +554,139 @@ func (n *NSQD) Notify(v interface{}) {
 	})
 }
 
+func uniqRands(l int, n int) []int {
+	set := make(map[int]struct{})
+	nums := make([]int, 0, l)
+	for {
+		num := rand.Intn(n)
+		if _, ok := set[num]; !ok {
+			set[num] = struct{}{}
+			nums = append(nums, num)
+		}
+		if len(nums) == l {
+			goto exit
+		}
+	}
+exit:
+	return nums
+}
+
+func (n *NSQD) resizePool(num int, workCh chan *Channel, responseCh chan bool, closeCh chan int) {
+	idealPoolSize := int(float64(num) * 0.25)
+	if idealPoolSize < 1 {
+		idealPoolSize = 1
+	} else if idealPoolSize > n.opts.QueueGCPoolMax {
+		idealPoolSize = n.opts.QueueGCPoolMax
+	}
+	for {
+		if idealPoolSize == n.poolSize {
+			break
+		} else if idealPoolSize < n.poolSize {
+			// contract
+			closeCh <- 1
+			n.poolSize--
+		} else {
+			// expand
+			n.waitGroup.Wrap(func() {
+				n.queueGCWorker(workCh, responseCh, closeCh)
+			})
+			n.poolSize++
+		}
+	}
+}
+
+func (n *NSQD) queueGCWorker(workCh chan *Channel, responseCh chan bool, closeCh chan int) {
+	for {
+		select {
+		case c := <-workCh:
+			now := time.Now().UnixNano()
+			dirty := false
+			if c.Exiting() {
+				responseCh <- dirty
+				continue
+			}
+			if c.processInFlightQueue(now) {
+				dirty = true
+			}
+			if c.processDeferredQueue(now) {
+				dirty = true
+			}
+			responseCh <- dirty
+		case <-closeCh:
+			return
+		case <-n.exitChan:
+			return
+		}
+	}
+}
+
+func (n *NSQD) channels() []*Channel {
+	var channels []*Channel
+	n.RLock()
+	for _, t := range n.topicMap {
+		t.RLock()
+		for _, c := range t.channelMap {
+			channels = append(channels, c)
+		}
+		t.RUnlock()
+	}
+	n.RUnlock()
+	return channels
+}
+
+func (n *NSQD) queueGCLoop() {
+	workCh := make(chan *Channel, n.opts.QueueGCSelectionCount)
+	responseCh := make(chan bool, n.opts.QueueGCSelectionCount)
+	closeCh := make(chan int)
+
+	workTicker := time.NewTicker(n.opts.QueueGCInterval)
+	refreshTicker := time.NewTicker(n.opts.QueueGCRefreshInterval)
+
+	channels := n.channels()
+	n.resizePool(len(channels), workCh, responseCh, closeCh)
+
+	for {
+		select {
+		case <-workTicker.C:
+			if len(channels) == 0 {
+				continue
+			}
+		case <-refreshTicker.C:
+			channels = n.channels()
+			n.resizePool(len(channels), workCh, responseCh, closeCh)
+			continue
+		case <-n.exitChan:
+			goto exit
+		}
+
+	loop:
+		num := n.opts.QueueGCSelectionCount
+		if num > len(channels) {
+			num = len(channels)
+		}
+
+		for _, i := range uniqRands(num, len(channels)) {
+			workCh <- channels[i]
+		}
+
+		numDirty := 0
+		for i := 0; i < num; i++ {
+			if <-responseCh {
```

Comment:
making a note for myself:

This is insufficient in the case where we asynchronously get a signal from `n.exitChan` to close... there's definitely a case where this would deadlock.

We should probably select on `n.exitChan` here, too.